src/workerd/api/streams/internal.c++

```diff
@@ -78,35 +78,117 @@ private:
+    // be lying about it's length, let's skip reading anything in this case.
     if (amountToRead > 0) {
       for (;;) {
-        // TODO(perf): We can likely further optimize this loop by checking to see
-        // how much of the buffer was filled and using the remaining buffer space if
-        // it is not completely filled by the previous iteration. Doing so makes this
-        // loop a bit more complicated tho, so for now let's keep things simple.
         auto bytes = kj::heapArray<T>(amountToRead);
-        size_t amount = co_await input.tryRead(bytes.begin(), 1, bytes.size());
+        // Note that we're passing amountToRead as the *minBytes* here so the tryRead should
+        // attempt to fill the entire buffer. If it doesn't, the implication is that we read
+        // everything.
+        uint64_t amount = co_await input.tryRead(bytes.begin(), amountToRead, amountToRead);
+        KJ_DASSERT(amount <= amountToRead);
 
-        if (amount == 0) {
+        runningTotal += amount;
+        JSG_REQUIRE(runningTotal < limit, TypeError, "Memory limit exceeded before EOF.");
+
+        if (amount < amountToRead) {
+          // The stream has indicated that we're all done by returning a value less than the
+          // full buffer length.
+          // It is possible/likely that at least some amount of data was written to the buffer.
+          // In which case we want to add that subset to the parts list here before we exit
+          // the loop.
+          if (amount > 0) {
+            parts.add(bytes.slice(0, amount).attach(kj::mv(bytes)));
+          }
           break;
         }
 
-        runningTotal += amount;
-        JSG_REQUIRE(runningTotal < limit, TypeError, "Memory limit exceeded before EOF.");
-        parts.add(bytes.slice(0, amount).attach(kj::mv(bytes)));
-      };
+        // Because we specify minSize equal to maxSize in the tryRead above, we should only
+        // get here if the buffer was completely filled by the read. If it wasn't completely
+        // filled, that is an indication that the stream is complete which is handled above.
+        KJ_DASSERT(amount == bytes.size());
+        parts.add(kj::mv(bytes));
+
+        // If the stream provided an expected length and our running total is equal to
+        // or greater than that length, let's use a potentially more conservative allocation
+        // for the next loop. In the best case scenario, the read will return 0 indicating
+        // that we're all done, in which case the allocation here is wasted so we want to
+        // be careful about not allocating too much.
+        // TODO(perf): Alternatively, we *could* choose to stop reading at this point if
+        // runningTotal >= length, taking the stream at it's word that the claimed length
+        // is all we're going to get. That would allow us to exit this loop now and avoid
+        // a potentially wasted additional allocation.
```

Comment:
@kentonv ... I think this would be the only remaining question. Should we just go ahead and take the stream at it's word if `tryGetLength(...)` returns a value and stop reading once we hit that? Doing so allows us to avoid the wasted additional allocation in this case.

Reply:
Yes, we can trust that `tryGetLength()` is accurate. FWIW the main implementation that returns non-null is for HTTP bodies with a content-length.